tests/libs/util/socket_helper.cpp

```diff
@@ -0,0 +1,326 @@
+    _In_z_ const char* message, sockaddr_storage& remote_address, uint16_t remote_port)
+{
+    // Send a message to the remote host using the sender socket.
+    ((PSOCKADDR_IN6)&remote_address)->sin6_port = htons(remote_port);
+    std::vector<char> send_buffer(message, message + strlen(message));
+    uint32_t bytes_sent = 0;
+    overlapped.hEvent = WSACreateEvent();
+    if (!connectex(
+            socket,
+            (PSOCKADDR)&remote_address,
+            sizeof(remote_address),
+            send_buffer.data(),
+            static_cast<DWORD>(send_buffer.size()),
+            reinterpret_cast<LPDWORD>(&bytes_sent),
+            &overlapped)) {
+        int wsaerr = WSAGetLastError();
+        if (wsaerr != WSA_IO_PENDING)
+            FAIL("ConnectEx failed with " << wsaerr);
+    }
+}
+
+void
+_stream_sender_socket::cancel_send_message()
+{
+    CancelIoEx((HANDLE)socket, &overlapped);
+    WSACloseEvent(overlapped.hEvent);
+}
+
+_receiver_socket::_receiver_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _base_socket{_sock_type, _protocol, _port}, overlapped{}, recv_buffer(std::vector<char>(1024)), recv_flags(0)
+{}
+
+_receiver_socket::~_receiver_socket() { WSACloseEvent(overlapped.hEvent); }
+
+void
+_receiver_socket::complete_async_receive(bool timeout_expected)
+{
+    int error = 0;
+    // Wait for the receiver socket to receive the message.
+    error = WSAWaitForMultipleEvents(1, &overlapped.hEvent, TRUE, 1000, TRUE);
+    if (error == WSA_WAIT_EVENT_0) {
+        if (timeout_expected)
+            FAIL("Receiver socket received a message when timeout was expected.");
+
+        if (!WSAGetOverlappedResult(
+                socket,
+                &overlapped,
+                reinterpret_cast<LPDWORD>(&bytes_received),
+                FALSE,
+                reinterpret_cast<LPDWORD>(&recv_flags)))
+            FAIL("WSArecvFrom on the receiver socket failed with error: " << WSAGetLastError());
+        WSACloseEvent(overlapped.hEvent);
+    } else {
+        if (error == WSA_WAIT_TIMEOUT) {
+            if (!timeout_expected)
+                FAIL("Receiver socket did not receive any message in 1 second.");
+        } else {
+            FAIL("Waiting on receiver socket failed with " << error);
+        }
+    }
+}
+
+void
+_receiver_socket::get_received_message(uint32_t& message_size, _Out_ char*& message)
+{
+    message_size = bytes_received;
+    message = recv_buffer.data();
+}
+
+_datagram_receiver_socket::_datagram_receiver_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _receiver_socket{_sock_type, _protocol, _port}, sender_address{}, sender_address_size(sizeof(sender_address))
+
+{
+    if (!(sock_type == SOCK_DGRAM || sock_type == SOCK_RAW) &&
+        !(protocol == IPPROTO_UDP || protocol == IPPROTO_IPV4 || protocol == IPPROTO_IPV6))
+        FAIL("datagram_socket only supports these combinations (SOCK_DGRAM, IPPROTO_UDP) and (SOCK_RAW, "
+             "IPPROTO_IPV4/IPV6)");
```

Comment:
doesn't match what 215-216 does, which also accept (SOCK_RAW, IPPROTO_UDP). and https://docs.microsoft.com/en-us/windows/win32/winsock/tcp-ip-raw-sockets-2 explicitly says windows supports that combo